src/goto-instrument/contracts/dynamic-frames/dfcc_cfg_info.cpp

```diff
@@ -0,0 +1,910 @@
+  local_may_aliast local_may_alias;
+  dirtyt dirty;
+  const namespacet ns;
+
+  bool is_assigned(const irep_idt &ident, assignst assigns)
+  {
+    PRECONDITION(!dirty(ident));
+    // For each assigns clause target
+    for(const auto &expr : assigns)
+    {
+      auto root_objects = dfcc_root_objects(expr);
+      for(const auto &root_object : root_objects)
+      {
+        if(
+          root_object.id() == ID_symbol &&
+          expr_try_dynamic_cast<symbol_exprt>(root_object)->get_identifier() ==
+            ident)
+        {
+          return true;
+        }
+        // If `root_object` is not a symbol, then it contains a combination of
+        // address-of and dereference operators that cannot be statically
+        // resolved to a symbol.
+        // Since we know `ident` is not dirty, we know that dereference
+        // operations cannot land on that `ident`. So the root_object cannot
+        // describe a memory location within the object backing that ident.
+        // We conclude that ident is not assigned by this target and move on to
+        // the next target.
+      }
+    }
+    return false;
+  }
+
+  void apply()
+  {
+    std::size_t top_level_id = loop_nesting_graph.size();
+
+    // initialize the DECL ident -> loop id map with top level id
+    for(auto target = goto_function.body.instructions.begin();
+        target != goto_function.body.instructions.end();
+        target++)
+    {
+      if(target->is_decl())
+      {
+        ident_to_loop_id_map[target->decl_symbol().get_identifier()] =
+          top_level_id;
+      }
+    }
+
+    for(std::size_t loop_id = 0; loop_id < loop_nesting_graph.size(); loop_id++)
+    {
+      // Recursive traversal from each top level loop
+      // (top level loops do not have successors)
+      if(loop_nesting_graph.get_successors(loop_id).empty())
+      {
+        top_level_loops.push_back(loop_id);
+        apply_rec(loop_id);
+      }
+    }
+
+    // Compute set of locals at function level
+    top_level_local.insert(
+      goto_function.parameter_identifiers.begin(),
+      goto_function.parameter_identifiers.end());
+
+    for(const auto &entry : ident_to_loop_id_map)
+    {
+      if(entry.second == loop_info_map.size())
+      {
+        top_level_local.insert(entry.first);
+      }
+    }
+
+    // Compute subset of function locals that must be tracked in the function's
+    // write set. A local must be tracked if it is dirty or if it may be
+    // assigned by one of the top level loops.
+    for(const auto &ident : top_level_local)
+    {
+      if(dirty(ident))
+      {
+        top_level_tracked.insert(ident);
+      }
+      else
+      {
+        // Check if this ident is touched by one of the top level loops
+        for(const auto loop_id : top_level_loops)
+        {
+          if(is_assigned(ident, loop_info_map.at(loop_id).assigns))
+          {
+            top_level_tracked.insert(ident);
+          }
```

Comment:
I think you should `break` after doing an `insert`, but more generally I am wondering whether this nesting of loops is the most efficient way of doing this. Need to build up a better understanding of the code to judge this.

Reply:
Yes we can break when we hit the first condition. 

What are you referring to by "this" in "this nesting of loops is the most efficient way of doing this" ?